govcd/provider_vdc.go

```diff
@@ -182,3 +185,405 @@ func getProviderVdcByName(vcdClient *VCDClient, providerVdcName string, extended
+			return fmt.Errorf("resource pool %s not found in provider VDC %s", rp.ResourcePool.Name, pvdc.VMWProviderVdc.Name)
+		}
+		if foundUsed.IsPrimary {
+			return fmt.Errorf("resource pool %s (%s) caannot be removed, because it is the primary one for provider VDC %s",
+				rp.ResourcePool.Name, rp.ResourcePool.Moref, pvdc.VMWProviderVdc.Name)
+		}
+		if foundUsed.IsEnabled {
+			err = disableResourcePool(pvdc.client, foundUsed.HREF)
+			if err != nil {
+				return fmt.Errorf("error disabling resource pool %s: %s", foundUsed.Name, err)
+			}
+		}
+
+		item := types.Reference{
+			HREF: foundUsed.HREF,
+			ID:   extractUuid(foundUsed.HREF),
+			Name: foundUsed.Name,
+			Type: "application/vnd.vmware.admin.vmwProviderVdcResourcePool+xml",
+		}
+		items = append(items, &item)
+	}
+
+	input := types.DeleteResourcePool{ResourcePoolRefs: items}
+
+	resp, err := pvdc.client.ExecuteJsonRequest(href, http.MethodPost, input, "error removing resource pools from provider VDC: %s")
+	if err != nil {
+		return err
+	}
+	task := NewTask(pvdc.client)
+	err = decodeBody(types.BodyTypeJSON, resp, task.Task)
+	if err != nil {
+		return err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return err
+	}
+	return pvdc.Refresh()
+}
+
+// GetResourcePools returns the Resource Pools belonging to this provider VDC
+func (pvdc *ProviderVdcExtended) GetResourcePools() ([]*types.QueryResultResourcePoolRecordType, error) {
+	resourcePools, err := pvdc.client.cumulativeQuery(types.QtResourcePool, map[string]string{
+		"type":          types.QtResourcePool,
+		"filter":        fmt.Sprintf("providerVdc==%s", url.QueryEscape(extractUuid(pvdc.VMWProviderVdc.HREF))),
+		"filterEncoded": "true",
+	}, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not get the Resource pool: %s", err)
+	}
+	return resourcePools.Results.ResourcePoolRecord, nil
+}
+
+// disableResourcePool disables a resource pool while it is assigned to a provider VDC
+// Calling this function is a prerequisite to removing a resource pool from a provider VDC
+func disableResourcePool(client *Client, resourcePoolHref string) error {
+	href, err := url.JoinPath(resourcePoolHref, "action", "disable")
+	if err != nil {
+		return err
+	}
+	return client.ExecuteRequestWithoutResponse(href, http.MethodPost, "", "error disabling resource pool: %s", nil)
+}
+
+// AddStorageProfiles adds the given storage profiles in this provider VDC
+func (pvdc *ProviderVdcExtended) AddStorageProfiles(storageProfileNames []string) error {
+	href, err := url.JoinPath(pvdc.VMWProviderVdc.HREF, "storageProfiles")
+	if err != nil {
+		return err
+	}
+
+	addStorageProfiles := &types.AddStorageProfiles{AddStorageProfile: storageProfileNames}
+
+	_, err = pvdc.client.ExecuteJsonRequest(href, http.MethodPost, addStorageProfiles,
+		"error adding storage profiles to provider VDC: %s")
+
+	if err != nil {
+		return err
+	}
+
+	err = pvdc.Refresh()
+	if err != nil {
+		return err
+	}
+
+	timeout := 30 * time.Second
```

Comment:
Probably not.
I am using this method because the return data structure is not immediately usable.
In fact, looking at the code I see the possibility of a bug. I'll try to unify the approach used in both `Update` and `AddStorageProfile`